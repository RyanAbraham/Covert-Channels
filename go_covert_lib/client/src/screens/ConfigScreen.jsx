import React from 'react';
import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import Dropdown from 'react-bootstrap/Dropdown';

import Checkbox from '../ui-components/Checkbox';
import IPInput from '../ui-components/IPInput';
import NumberInput from '../ui-components/NumberInput';
import Select from '../ui-components/Select';
import TextArea from '../ui-components/TextArea';

const ConfigScreen = (props) => {
  const {
    openChannel,
    closeChannel,
    config,
    setConfig,
    processorList,
    processors,
    setProcessors,
    channelList,
    channel,
    setChannel,
    channelIsOpen,
  } = props;

  return (
    <div className="m-2">
      <h2 className="m-1">Configuration</h2>
      <h3 className="m-1">Processors</h3>
      <Button
        variant="success"
        className="m-1 w-100"
        onClick={() => setProcessors(processors.concat({
          Type: null,
          Data: null,
        }))}
      >
        Add Processor
      </Button>
      {
        processors.map((processor, i) => (
          <div key={i.toString()}>
            <Dropdown className="m-1">
              <Dropdown.Toggle
                className="w-100"
                variant="outline-primary"
              >
                {processor.Type || 'Select a Processor'}
              </Dropdown.Toggle>
              <Dropdown.Menu className="w-100">
                {
                  Object.keys(processorList).map(p => (
                    <Dropdown.Item
                      as="option"
                      active={p === processor.Type}
                      onClick={(e) => {
                        setProcessors([
                          ...processors.slice(0, i),
                          {
                            Type: e.target.value,
                            Data: processorList,
                          },
                          ...processors.slice(i + 1, processors.length + 1),
                        ]);
                      }}
                      value={p}
                      key={p}
                    >
                      {p}
                    </Dropdown.Item>
                  ))
                }
              </Dropdown.Menu>
            </Dropdown>
            {processor.Data && Object.keys(processor.Data[processor.Type]).map((key) => {
              /**
               * EXTREME DANGER WARNING
               * The below code involves very convoluted spread operators to massage
               * the data to the format that GoLang expects it to be in.
               */
              const opt = processor.Data[processor.Type][key];
              let propsForComponent = {
                key,
                label: opt.Display.Name,
                value: opt.Value,
                tooltip: opt.Display.Description,
                parentOnChange: e => setProcessors([
                  ...processors.slice(0, i),
                  {
                    ...processor,
                    Data: {
                      ...processor.Data,
                      [processor.Type]: {
                        ...processor.Data[processor.Type],
                        [key]: {
                          ...opt,
                          Value: e.target.value,
                        },
                      },
                    },
                  },
                  ...processors.slice(i + 1, processors.length + 1),
                ]),
              };
              switch (opt.Type) {
                case 'ipv4':
                  return (
                    <IPInput {...propsForComponent} />
                  );
                case 'i8':
                case 'u16':
                case 'u64':
                  if (opt.Range) {
                    propsForComponent = {
                      ...propsForComponent,
                      min: opt.Range[0],
                      max: opt.Range[1],
                    };
                  }
                  return (
                    <NumberInput
                      {...propsForComponent}
                      parentOnChange={e => setProcessors([
                        ...processors.slice(0, i),
                        {
                          ...processor,
                          Data: {
                            ...processor.Data,
                            [processor.Type]: {
                              ...processor.Data[processor.Type],
                              [key]: {
                                ...opt,
                                Value: parseInt(e.target.value) || 0,
                              },
                            },
                          },
                        },
                        ...processors.slice(i + 1, processors.length + 1),
                      ])}
                    />
                  );
                case 'exactu64':
                  return (<div>EXACTU64</div>);
                case 'bool':
                  return (
                    <Checkbox
                      {...propsForComponent}
                      parentOnChange={e => setProcessors([
                        ...processors.slice(0, i),
                        {
                          ...processor,
                          Data: {
                            ...processor.Data,
                            [processor.Type]: {
                              ...processor.Data[processor.Type],
                              [key]: {
                                ...opt,
                                Value: e.target.checked,
                              },
                            },
                          },
                        },
                        ...processors.slice(i + 1, processors.length + 1),
                      ])}
                    />
                  );
                case 'select':
                  return (
                    <Select
                      {...propsForComponent}
                      items={opt.Range}
                    />
                  );
                case 'hexkey':
                  return (<div>hexkey</div>);
                case 'key':
                  return (<TextArea {...propsForComponent} />);
                default:
                  return (<div key={key}>UNIMPLEMENTED</div>);
              }
            })}
          </div>
        ))
      }
      <h3 className="m-1">Channel</h3>
      <Dropdown className="m-1">
        <Dropdown.Toggle
          className="cc-config__chan-select w-100"
          variant="outline-primary"
        >
          {channel.value || 'Select a Channel'}
        </Dropdown.Toggle>
        <Dropdown.Menu className="w-100">
          {
            Object.keys(channelList).map(chan => (
              <Dropdown.Item
                as="option"
                active={chan === channel.value}
                onClick={(e) => {
                  setChannel({
                    value: e.target.value,
                    properties: channelList[chan],
                  });
                  setConfig(channelList[chan]);
                }}
                value={chan}
                key={chan}
              >
                {chan}
              </Dropdown.Item>
            ))
          }
        </Dropdown.Menu>
      </Dropdown>
      {Object.keys(config).map((key) => {
        const opt = config[key];
        let propsForComponent = {
          key,
          label: opt.Display.Name,
          value: opt.Value,
          tooltip: opt.Display.Description,
          parentOnChange: e => setConfig({
            ...config,
            [key]: {
              ...config[key],
              Value: e.target.value,
            },
          }),
        };
        switch (opt.Type) {
          case 'ipv4':
            return (
              <IPInput {...propsForComponent} />
            );
          case 'i8':
          case 'u16':
          case 'u64':
            if (opt.Range) {
              propsForComponent = {
                ...propsForComponent,
                min: opt.Range[0],
                max: opt.Range[1],
              };
            }
            return (
              <NumberInput
                {...propsForComponent}
                parentOnChange={e => setConfig({
                  ...config,
                  [key]: {
                    ...config[key],
                    Value: parseInt(e.target.value) || 0,
                  },
                })}
              />
            );
          case 'exactu64':
            return (<div>EXACTU64</div>);
          case 'bool':
            return (
              <Checkbox
                {...propsForComponent}
                parentOnChange={e => setConfig({
                  ...config,
                  [key]: {
                    ...config[key],
                    Value: e.target.checked,
                  },
                })}
              />
            );
          case 'select':
            return (
              <Select
                {...propsForComponent}
                items={opt.Range}
              />
            );
          case 'hexkey':
            return (<div>hexkey</div>);
          case 'key':
            return (<TextArea {...propsForComponent} />);
          default:
            return (<div key={key}>UNIMPLEMENTED</div>);
        }
      })}
      {channelIsOpen ? (
        <Button variant="danger" onClick={closeChannel} className="m-1 w-100">Close Covert Channel</Button>
      ) : (
        <Button
          variant="success"
          onClick={openChannel}
          className="cc-config__submit m-1 w-100"
          hidden={Object.entries(channel).length === 0 && channel.constructor === Object}
        >
          Open Covert Channel
        </Button>
      )}
    </div>
  );
};

ConfigScreen.propTypes = {
  openChannel: PropTypes.func.isRequired,
  closeChannel: PropTypes.func.isRequired,
  config: PropTypes.object.isRequired,
  setConfig: PropTypes.func.isRequired,
  processorList: PropTypes.object.isRequired,
  processors: PropTypes.array.isRequired,
  setProcessors: PropTypes.func.isRequired,
  channelList: PropTypes.object.isRequired,
  channel: PropTypes.object.isRequired,
  setChannel: PropTypes.func.isRequired,
  channelIsOpen: PropTypes.bool.isRequired,
};

export default ConfigScreen;
