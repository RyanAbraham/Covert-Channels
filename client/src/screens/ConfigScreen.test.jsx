/* eslint-disable no-use-before-define */
import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import ConfigScreen from './ConfigScreen';

const sampleProcessorList = {
  Caesar: {
    Shift: {
      Type: 'i8',
      Value: 49,
      Range: [
        -128,
        127,
      ],
      Display: {
        Description: 'The shift for the Caesar cypher',
        Name: 'Shift',
        Group: '',
        GroupToggle: false,
      },
    },
  },
  Checksum: {
    Polynomial: {
      Type: 'select',
      Value: 'IEEE',
      Range: [
        'IEEE',
        'Castagnoli',
        'Koopman',
      ],
      Display: {
        Description: 'The predefined polynomial to use for the crc32 checksum',
        Name: '',
        Group: '',
        GroupToggle: false,
      },
    },
  },
};

const sampleChannelList = {
  TcpSyn: {
    FriendIP: {
      Type: 'ipv4',
      Value: '127.0.0.1',
      Display: {
        Description: 'Your friend\'s IP address.',
        Name: 'Friend\'s IP',
        Group: 'IP Addresses',
        GroupToggle: false,
      },
    },
    OriginIP: {
      Type: 'ipv4',
      Value: '127.0.0.1',
      Display: {
        Description: 'Your IP address.',
        Name: 'Your IP',
        Group: 'IP Addresses',
        GroupToggle: false,
      },
    },
  },
  TcpHandshake: {},
};

// Polyfill which mocks document functions needed for Bootstrap components
global.document.createRange = () => ({
  setStart: () => {},
  setEnd: () => {},
  commonAncestorContainer: {
    nodeName: 'BODY',
    ownerDocument: document,
  },
});

describe('Config Screen', () => {
  let cfgScreen;
  const openChannel = jest.fn();
  const closeChannel = jest.fn();
  const config = {};
  const setConfig = jest.fn();
  const processorList = sampleProcessorList;
  const processors = [];
  const setProcessors = jest.fn();
  const channelList = sampleChannelList;
  const channel = {};
  const setChannel = jest.fn(() => cfgScreen.setProps({
    config: channelList.TcpSyn,
    channel: {
      value: 'TcpSyn',
      properties: channelList.TcpSyn,
    },
  }));
  const channelIsOpen = false;

  beforeEach(() => {
    cfgScreen = mount(<ConfigScreen
      openChannel={openChannel}
      closeChannel={closeChannel}
      config={config}
      setConfig={setConfig}
      processorList={processorList}
      processors={processors}
      setProcessors={setProcessors}
      channelList={channelList}
      channel={channel}
      setChannel={setChannel}
      channelIsOpen={channelIsOpen}
    />);
  });

  afterEach(() => {
    act(() => {
      cfgScreen.unmount();
    });
  });

  it('displays a configuration object correctly', async () => {
    expect(cfgScreen.debug()).toMatchSnapshot();
    await act(async () => {
      const chanSelect = cfgScreen.find('.cc-config__chan-select').at(0);
      chanSelect.simulate('click');
    });
    cfgScreen.update(); // Need to update because Bootstrap dropdowns are fancy and finicky
    const firstDropdownOption = cfgScreen.find('option').at(0);
    firstDropdownOption.simulate('click');
    expect(setChannel).toHaveBeenCalledTimes(1);
    expect(cfgScreen.debug()).toMatchSnapshot();
  });

  it('submits a configuration correctly', () => {
    setChannel();
    const submitButton = cfgScreen.find('.cc-config__submit').at(0);
    submitButton.simulate('click');
    expect(openChannel).toHaveBeenCalledTimes(1);
    expect(cfgScreen.debug()).toMatchSnapshot();
  });
});
