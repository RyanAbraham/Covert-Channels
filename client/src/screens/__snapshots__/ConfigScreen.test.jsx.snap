// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Screen displays a configuration object correctly 1`] = `
"<ConfigScreen openChannel={[Function: mockConstructor]} closeChannel={[Function: mockConstructor]} config={{...}} setConfig={[Function: mockConstructor]} processorList={{...}} processors={{...}} setProcessors={[Function: mockConstructor]} channelList={{...}} channel={{...}} setChannel={[Function: mockConstructor]} channelIsOpen={false}>
  <div className=\\"m-2\\">
    <h2 className=\\"m-1\\">
      Configuration
    </h2>
    <h3 className=\\"m-1\\">
      Processors
    </h3>
    <Button variant=\\"success\\" className=\\"m-1 w-100\\" onClick={[Function: onClick]} active={false} disabled={false} type=\\"button\\">
      <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"m-1 w-100 btn btn-success\\">
        Add Processor
      </button>
    </Button>
    <h3 className=\\"m-1\\">
      Channel
    </h3>
    <Dropdown className=\\"m-1\\" navbar={false}>
      <ReactOverlaysDropdown drop={[undefined]} show={[undefined]} alignEnd={[undefined]} onToggle={[Function]} focusFirstItemOnShow={[undefined]} itemSelector=\\".dropdown-item:not(.disabled):not(:disabled)\\">
        <div onKeyDown={[Function: handleKeyDown]} className=\\"m-1 dropdown\\">
          <DropdownToggle className=\\"cc-config__chan-select w-100\\" variant=\\"outline-primary\\">
            <Button onClick={[Function]} className=\\"cc-config__chan-select w-100 dropdown-toggle\\" aria-haspopup={true} aria-expanded={false} variant=\\"outline-primary\\" active={false} disabled={false} type=\\"button\\">
              <button onClick={[Function]} aria-haspopup={true} aria-expanded={false} disabled={false} type=\\"button\\" className=\\"cc-config__chan-select w-100 dropdown-toggle btn btn-outline-primary\\">
                Select a Channel
              </button>
            </Button>
          </DropdownToggle>
          <DropdownMenu className=\\"w-100\\" alignRight={false} flip={true} />
        </div>
      </ReactOverlaysDropdown>
    </Dropdown>
    <Button variant=\\"success\\" onClick={[Function: mockConstructor]} className=\\"cc-config__submit m-1 w-100\\" hidden={true} active={false} disabled={false} type=\\"button\\">
      <button onClick={[Function: mockConstructor]} hidden={true} disabled={false} type=\\"button\\" className=\\"cc-config__submit m-1 w-100 btn btn-success\\">
        Open Covert Channel
      </button>
    </Button>
  </div>
</ConfigScreen>"
`;

exports[`Config Screen displays a configuration object correctly 2`] = `
"<ConfigScreen openChannel={[Function: mockConstructor]} closeChannel={[Function: mockConstructor]} config={{...}} setConfig={[Function: mockConstructor]} processorList={{...}} processors={{...}} setProcessors={[Function: mockConstructor]} channelList={{...}} channel={{...}} setChannel={[Function: mockConstructor]} channelIsOpen={false}>
  <div className=\\"m-2\\">
    <h2 className=\\"m-1\\">
      Configuration
    </h2>
    <h3 className=\\"m-1\\">
      Processors
    </h3>
    <Button variant=\\"success\\" className=\\"m-1 w-100\\" onClick={[Function: onClick]} active={false} disabled={false} type=\\"button\\">
      <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"m-1 w-100 btn btn-success\\">
        Add Processor
      </button>
    </Button>
    <h3 className=\\"m-1\\">
      Channel
    </h3>
    <Dropdown className=\\"m-1\\" navbar={false}>
      <ReactOverlaysDropdown drop={[undefined]} show={false} alignEnd={[undefined]} onToggle={[Function]} focusFirstItemOnShow={[undefined]} itemSelector=\\".dropdown-item:not(.disabled):not(:disabled)\\">
        <div onKeyDown={[Function: handleKeyDown]} className=\\"m-1 dropdown\\">
          <DropdownToggle className=\\"cc-config__chan-select w-100\\" variant=\\"outline-primary\\">
            <Button onClick={[Function]} className=\\"cc-config__chan-select w-100 dropdown-toggle\\" aria-haspopup={true} aria-expanded={false} variant=\\"outline-primary\\" active={false} disabled={false} type=\\"button\\">
              <button onClick={[Function]} aria-haspopup={true} aria-expanded={false} disabled={false} type=\\"button\\" className=\\"cc-config__chan-select w-100 dropdown-toggle btn btn-outline-primary\\">
                TcpSyn
              </button>
            </Button>
          </DropdownToggle>
          <DropdownMenu className=\\"w-100\\" alignRight={false} flip={true}>
            <div x-placement=\\"bottom-start\\" aria-labelledby=\\"\\" style={{...}} className=\\"w-100 dropdown-menu\\">
              <DropdownItem as=\\"option\\" active={true} onClick={[Function: onClick]} value=\\"TcpSyn\\" disabled={false}>
                <option value=\\"TcpSyn\\" href={[undefined]} disabled={false} className=\\"dropdown-item active\\" onClick={[Function]}>
                  TcpSyn
                </option>
              </DropdownItem>
              <DropdownItem as=\\"option\\" active={false} onClick={[Function: onClick]} value=\\"TcpHandshake\\" disabled={false}>
                <option value=\\"TcpHandshake\\" href={[undefined]} disabled={false} className=\\"dropdown-item\\" onClick={[Function]}>
                  TcpHandshake
                </option>
              </DropdownItem>
            </div>
          </DropdownMenu>
        </div>
      </ReactOverlaysDropdown>
    </Dropdown>
    <IPInput label=\\"Friend\\\\'s IP\\" value=\\"127.0.0.1\\" tooltip=\\"Your friend\\\\'s IP address.\\" parentOnChange={[Function: parentOnChange]}>
      <div>
        <InputGroup className=\\"cc-ip-input m-1 w-100 \\">
          <div className=\\"cc-ip-input m-1 w-100  input-group\\">
            <InputGroupPrepend>
              <div className=\\"input-group-prepend\\">
                <InputGroupText className=\\"input-text\\">
                  <span className=\\"input-text input-group-text\\">
                    Friend&#39;s IP
                  </span>
                </InputGroupText>
              </div>
            </InputGroupPrepend>
            <FormControl value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]}>
              <input value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]} type={[undefined]} readOnly={[undefined]} id={[undefined]} className=\\"form-control\\" />
            </FormControl>
            <OverlayTrigger overlay={{...}} defaultOverlayShown={false} trigger={{...}}>
              <RefHolder>
                <span className=\\"cc-tooltip ml-1 mr-1\\" onFocus={[Function]} onBlur={[Function]} onMouseOver={[Function]} onMouseOut={[Function]}>
                  <div className=\\"cc-tooltip__icon\\">
                    ?
                  </div>
                </span>
              </RefHolder>
              <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} transition={{...}} rootClose={false} placement=\\"top\\">
                <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} rootClose={false} placement=\\"top\\" transition={{...}} containerPadding={5} />
              </Overlay>
            </OverlayTrigger>
          </div>
        </InputGroup>
      </div>
    </IPInput>
    <IPInput label=\\"Your IP\\" value=\\"127.0.0.1\\" tooltip=\\"Your IP address.\\" parentOnChange={[Function: parentOnChange]}>
      <div>
        <InputGroup className=\\"cc-ip-input m-1 w-100 \\">
          <div className=\\"cc-ip-input m-1 w-100  input-group\\">
            <InputGroupPrepend>
              <div className=\\"input-group-prepend\\">
                <InputGroupText className=\\"input-text\\">
                  <span className=\\"input-text input-group-text\\">
                    Your IP
                  </span>
                </InputGroupText>
              </div>
            </InputGroupPrepend>
            <FormControl value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]}>
              <input value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]} type={[undefined]} readOnly={[undefined]} id={[undefined]} className=\\"form-control\\" />
            </FormControl>
            <OverlayTrigger overlay={{...}} defaultOverlayShown={false} trigger={{...}}>
              <RefHolder>
                <span className=\\"cc-tooltip ml-1 mr-1\\" onFocus={[Function]} onBlur={[Function]} onMouseOver={[Function]} onMouseOut={[Function]}>
                  <div className=\\"cc-tooltip__icon\\">
                    ?
                  </div>
                </span>
              </RefHolder>
              <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} transition={{...}} rootClose={false} placement=\\"top\\">
                <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} rootClose={false} placement=\\"top\\" transition={{...}} containerPadding={5} />
              </Overlay>
            </OverlayTrigger>
          </div>
        </InputGroup>
      </div>
    </IPInput>
    <Button variant=\\"success\\" onClick={[Function: mockConstructor]} className=\\"cc-config__submit m-1 w-100\\" hidden={false} active={false} disabled={false} type=\\"button\\">
      <button onClick={[Function: mockConstructor]} hidden={false} disabled={false} type=\\"button\\" className=\\"cc-config__submit m-1 w-100 btn btn-success\\">
        Open Covert Channel
      </button>
    </Button>
  </div>
</ConfigScreen>"
`;

exports[`Config Screen submits a configuration correctly 1`] = `
"<ConfigScreen openChannel={[Function: mockConstructor]} closeChannel={[Function: mockConstructor]} config={{...}} setConfig={[Function: mockConstructor]} processorList={{...}} processors={{...}} setProcessors={[Function: mockConstructor]} channelList={{...}} channel={{...}} setChannel={[Function: mockConstructor]} channelIsOpen={false}>
  <div className=\\"m-2\\">
    <h2 className=\\"m-1\\">
      Configuration
    </h2>
    <h3 className=\\"m-1\\">
      Processors
    </h3>
    <Button variant=\\"success\\" className=\\"m-1 w-100\\" onClick={[Function: onClick]} active={false} disabled={false} type=\\"button\\">
      <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"m-1 w-100 btn btn-success\\">
        Add Processor
      </button>
    </Button>
    <h3 className=\\"m-1\\">
      Channel
    </h3>
    <Dropdown className=\\"m-1\\" navbar={false}>
      <ReactOverlaysDropdown drop={[undefined]} show={[undefined]} alignEnd={[undefined]} onToggle={[Function]} focusFirstItemOnShow={[undefined]} itemSelector=\\".dropdown-item:not(.disabled):not(:disabled)\\">
        <div onKeyDown={[Function: handleKeyDown]} className=\\"m-1 dropdown\\">
          <DropdownToggle className=\\"cc-config__chan-select w-100\\" variant=\\"outline-primary\\">
            <Button onClick={[Function]} className=\\"cc-config__chan-select w-100 dropdown-toggle\\" aria-haspopup={true} aria-expanded={false} variant=\\"outline-primary\\" active={false} disabled={false} type=\\"button\\">
              <button onClick={[Function]} aria-haspopup={true} aria-expanded={false} disabled={false} type=\\"button\\" className=\\"cc-config__chan-select w-100 dropdown-toggle btn btn-outline-primary\\">
                TcpSyn
              </button>
            </Button>
          </DropdownToggle>
          <DropdownMenu className=\\"w-100\\" alignRight={false} flip={true} />
        </div>
      </ReactOverlaysDropdown>
    </Dropdown>
    <IPInput label=\\"Friend\\\\'s IP\\" value=\\"127.0.0.1\\" tooltip=\\"Your friend\\\\'s IP address.\\" parentOnChange={[Function: parentOnChange]}>
      <div>
        <InputGroup className=\\"cc-ip-input m-1 w-100 \\">
          <div className=\\"cc-ip-input m-1 w-100  input-group\\">
            <InputGroupPrepend>
              <div className=\\"input-group-prepend\\">
                <InputGroupText className=\\"input-text\\">
                  <span className=\\"input-text input-group-text\\">
                    Friend&#39;s IP
                  </span>
                </InputGroupText>
              </div>
            </InputGroupPrepend>
            <FormControl value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]}>
              <input value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]} type={[undefined]} readOnly={[undefined]} id={[undefined]} className=\\"form-control\\" />
            </FormControl>
            <OverlayTrigger overlay={{...}} defaultOverlayShown={false} trigger={{...}}>
              <RefHolder>
                <span className=\\"cc-tooltip ml-1 mr-1\\" onFocus={[Function]} onBlur={[Function]} onMouseOver={[Function]} onMouseOut={[Function]}>
                  <div className=\\"cc-tooltip__icon\\">
                    ?
                  </div>
                </span>
              </RefHolder>
              <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} transition={{...}} rootClose={false} placement=\\"top\\">
                <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} rootClose={false} placement=\\"top\\" transition={{...}} containerPadding={5} />
              </Overlay>
            </OverlayTrigger>
          </div>
        </InputGroup>
      </div>
    </IPInput>
    <IPInput label=\\"Your IP\\" value=\\"127.0.0.1\\" tooltip=\\"Your IP address.\\" parentOnChange={[Function: parentOnChange]}>
      <div>
        <InputGroup className=\\"cc-ip-input m-1 w-100 \\">
          <div className=\\"cc-ip-input m-1 w-100  input-group\\">
            <InputGroupPrepend>
              <div className=\\"input-group-prepend\\">
                <InputGroupText className=\\"input-text\\">
                  <span className=\\"input-text input-group-text\\">
                    Your IP
                  </span>
                </InputGroupText>
              </div>
            </InputGroupPrepend>
            <FormControl value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]}>
              <input value=\\"127.0.0.1\\" onChange={[Function: onChange]} onBlur={[Function: onBlur]} type={[undefined]} readOnly={[undefined]} id={[undefined]} className=\\"form-control\\" />
            </FormControl>
            <OverlayTrigger overlay={{...}} defaultOverlayShown={false} trigger={{...}}>
              <RefHolder>
                <span className=\\"cc-tooltip ml-1 mr-1\\" onFocus={[Function]} onBlur={[Function]} onMouseOver={[Function]} onMouseOut={[Function]}>
                  <div className=\\"cc-tooltip__icon\\">
                    ?
                  </div>
                </span>
              </RefHolder>
              <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} transition={{...}} rootClose={false} placement=\\"top\\">
                <Overlay defaultOverlayShown={false} popperConfig={{...}} show={false} onHide={[Function]} target={[Function]} rootClose={false} placement=\\"top\\" transition={{...}} containerPadding={5} />
              </Overlay>
            </OverlayTrigger>
          </div>
        </InputGroup>
      </div>
    </IPInput>
    <Button variant=\\"success\\" onClick={[Function: mockConstructor]} className=\\"cc-config__submit m-1 w-100\\" hidden={false} active={false} disabled={false} type=\\"button\\">
      <button onClick={[Function: mockConstructor]} hidden={false} disabled={false} type=\\"button\\" className=\\"cc-config__submit m-1 w-100 btn btn-success\\">
        Open Covert Channel
      </button>
    </Button>
  </div>
</ConfigScreen>"
`;
